Add equality comparison to c10::Dict

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35019 [jit] kill isSameIdentity
* #35005 [jit] kill `shallowEquals`
* #34986 [jit] define equality for IValue
* #34985 Move ivalue tests to aten
* **#34892 Add equality comparison to c10::Dict**
* #34856 Add equality comparisons to c10::List

Same rationale and implementation as in
https://github.com/pytorch/pytorch/pull/34856

Differential Revision: [D20493169](https://our.internmc.facebook.com/intern/diff/D20493169)

suo committed:
Add equality comparison to c10::Dict

Same rationale and implementation as in
https://github.com/pytorch/pytorch/pull/34856

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/Dict.h:
```diff
@@ -338,6 +338,25 @@ class Dict final {
+      const Dict<Key_, Value_>& lhs,
+      const Dict<Key_, Value_>& rhs);
+  template <class Key_, class Value_>
+  friend bool operator!=(
```
notice `!=` doesn't have an impl or test, maybe that's intended?

suo replied:
good catch! will add

bhosmer reviewed (approved):
Couple comments from the peanut gallery FWIW

suo commented on aten/src/ATen/core/Dict.h:
```diff
@@ -338,6 +338,25 @@ class Dict final {
+   * Identity comparison. Returns true if and only if `rhs` represents the same
+   * Dict object as `this`.
+   */
+  bool is(const Dict& is) const;
```
whoops lol

suo committed:
Update on "Add equality comparison to c10::Dict"


Same rationale and implementation as in
https://github.com/pytorch/pytorch/pull/34856

Differential Revision: [D20493169](https://our.internmc.facebook.com/intern/diff/D20493169)

[ghstack-poisoned]

suo committed:
Update on "Add equality comparison to c10::Dict"


Same rationale and implementation as in
https://github.com/pytorch/pytorch/pull/34856

Differential Revision: [D20493169](https://our.internmc.facebook.com/intern/diff/D20493169)

[ghstack-poisoned]
